src/compose/ComposeBox.js

```diff
@@ -184,6 +195,64 @@ class ComposeBox extends PureComponent<Props, State> {
     dispatch(draftUpdate(narrow, message));
   };
 
+  handleMentionSubscribedCheck = async (message: string) => {
+    const { usersByEmail, narrow, auth, streamId } = this.props;
+
+    if (isPrivateNarrow(narrow)) {
+      return;
+    }
+    const unformattedMessage = message.split('**')[1];
+
+    // We skip user groups, for which autocompletes are of the form
+    // `*<user_group_name>*`, and therefore, message.split('**')[1]
+    // is undefined.
+    if (unformattedMessage === undefined) {
+      return;
+    }
+    const [userFullName, userId] = unformattedMessage.split('|');
+    const unsubscribedMentions = this.state.unsubscribedMentions.slice();
+    let mentionedUser: UserOrBot;
+
+    // eslint-disable-next-line no-unused-vars
+    for (const [email, user] of usersByEmail) {
+      if (userId !== undefined) {
+        if (user.user_id === userId) {
+          mentionedUser = user;
+          break;
+        }
+      } else if (user.full_name === userFullName) {
+        mentionedUser = user;
+        break;
+      }
+    }
+    if (!mentionedUser || unsubscribedMentions.includes(mentionedUser)) {
+      return;
+    }
+
+    if (!(await api.getSubscriptionToStream(auth, mentionedUser.user_id, streamId)).is_subscribed) {
+      unsubscribedMentions.push(mentionedUser.user_id);
+      this.setState({ unsubscribedMentions });
+    }
+  };
+
+  handleMentionWarningDismiss = (user: UserOrBot) => {
+    this.setState(prevState => ({
+      unsubscribedMentions: prevState.unsubscribedMentions.filter(
+        (x: number) => x !== user.user_id,
+      ),
+    }));
+  };
+
+  clearMentionWarnings = () => {
+    this.setState({ unsubscribedMentions: [] });
+  };
```

Comment:
The `ComposeBox` component already handles a ton of different things, and that's already made it a lot more challenging to understand than it ought to be. So I'd like to avoid adding more, and instead put new functionality in appropriate new components and  have `ComposeBox` just do the minimum to wire them in.

Here, I think a good way to do that would be to have a component that
* has zero or more individual `MentionedUserNotSubscribed` components as children
* has the `unsubscribedMentions` state
* has the implementations of these three methods

Then from `ComposeBox` we just need to be able to invoke `handleMentionSubscribedCheck` and `clearMentionWarnings`. I think the React way to do this is to have `ComposeBox` hold a "ref":
https://reactjs.org/docs/refs-and-the-dom.html
using `createRef`, that points to that child component.

(Alternatively the state could live up at `ComposeBox`, and it could just pass the child some callbacks to get at it. But that would expose somewhat more of the details to `ComposeBox`, and because of the situation where `ComposeBox` has too many details of too many things already, I'd rather avoid that.)

Reply:
Done. The new component is `MentionWarnings`.